records/2024-08-14.suspend-resume-callbacks.md

```diff
@@ -0,0 +1,120 @@
+3. GCNotify sends an [email](https://github.com/cds-snc/notification-api/pull/2190/files#diff-a7de7566620064ea85eacda53163429be50282374a124c93fc04bc5f637a5dcbR29) to the service owner, making them aware of issues with their API.
+4. Resume the service's callbacks, via the platform admin UI, once the service owner contacts us
+
+#### Visualized flow
+```mermaid
+flowchart TB
+START((start)) --> A
+A{Did Callbacks\nrequire retries?}
+
+A == Yes ==> A1[Send warning email] ==> B{Issue\nresolved?}
+A -. No .-> B1
+
+B -- Yes --> B1((END))
+B == No ==> D{Callbacks reaching\nmax retries?}
+
+D == Yes ==> DYes1[Trigger critical alarm] ==> DYes2[Manually suspend callbacks] ==> DYes3[Email service about suspension] ==> E{Has service owner\ncontacted us?}
+D -. No .-> F{24h elapsed?}
+
+E == Yes ==> EY1[Manually resume callbacks] --> EY2((END))
+E -. Wait .-> E
+
+F -. No .-> B
+F -. Yes .-> A
+
+style START fill::#bbf,stroke:#fff,stroke-width:2px,color:#fff,stroke-dasharray: 5 5
+style B1 fill::#,stroke:#E35,stroke-width:2px,color:#fff,stroke-dasharray: 5 5
+style EY2 fill::#bbf,stroke:#E35,stroke-width:2px,color:#fff,stroke-dasharray: 5 5
+
+```
+
+### Implementation overview
+To keep track of a service's suspension status and when the suspension took place we will add two new columns to the `service_callback_api` table:
+
+- `suspended`
+- `suspended_at`
+
+When the platform admin UI is used to suspend a service, we will set `suspended` to `True` and record the time in the `suspended_at` column. By storing these values we facilitate tracking of suspended services so they can be easily displayed in the platform admin UI to be quickly resumed after the service resolves the issues with their API and contacts us.
+
+Should we automate the suspension and resumption process later down the road, the `suspended_at` column will provided us with the context we need to determine when we can automatically resume a service's callbacks. (e.g Time has reached a pre-defined threshold and we resume their callbacks.)
+
+## Automated suspension and resumption of service callbacks
+
+
+## Additional considerations
+
+### The burden of manual intervention
+As of now the frequency at which callback API's go down during larger sends, leading strain on the system, is unknown. Should this be a rare occurrence, then manual intervention should suffice as a course of action.
+
+However if this issue occurs more frequntly, particularly as Notify grows, then we should consider automating the suspension and resumption process in order to eliminate the burden of manual intervention, and maintain our velocity as a team.
+
+## Future considerations
+### Maintaining data quality as Notify grows
+Completeness, consistency, reliability and timeliness are pillars of data quality. While service owners are responsible for maintaining uptime of their API, in the case of their service going down, it's fair to assume the onus is on them to improve their uptime.
```

Comment:
excellent point and yes if we chose to automate this in the future I believe your current design will allow us to expand and do so